db/hints/manager.cc

```diff
@@ -223,76 +234,83 @@ future<> manager::wait_for_sync_point(abort_source& as, const sync_point::shard_
     co_return;
 }
 
-hint_endpoint_manager& manager::get_ep_manager(endpoint_id ep) {
-    auto it = find_ep_manager(ep);
-    if (it == ep_managers_end()) {
-        manager_logger.trace("Creating an ep_manager for {}", ep);
-        hint_endpoint_manager& ep_man = _ep_managers.emplace(ep, hint_endpoint_manager(ep, *this)).first->second;
-        ep_man.start();
-        return ep_man;
+hint_host_manager& manager::get_host_manager(locator::host_id hid) {
+    auto it = find_host_manager(hid);
+
+    if (it == host_managers_end()) {
+        manager_logger.trace("Creating an ep_manager for {}", hid);
+
+        hint_host_manager& hh_man = _host_managers.emplace(hid, hint_host_manager(hid, *this)).first->second;
+        hh_man.start();
+        
+        return hh_man;
     }
+
     return it->second;
 }
 
-inline bool manager::have_ep_manager(endpoint_id ep) const noexcept {
-    return find_ep_manager(ep) != ep_managers_end();
+inline bool manager::have_host_manager(locator::host_id hid) const noexcept {
+    return find_host_manager(hid) != host_managers_end();
 }
 
-bool manager::store_hint(endpoint_id ep, schema_ptr s, lw_shared_ptr<const frozen_mutation> fm, tracing::trace_state_ptr tr_state) noexcept {
-    if (stopping() || draining_all() || !started() || !can_hint_for(ep)) {
-        manager_logger.trace("Can't store a hint to {}", ep);
+bool manager::store_hint(locator::host_id hid, schema_ptr s, lw_shared_ptr<const frozen_mutation> fm, tracing::trace_state_ptr tr_state) noexcept {
+    if (stopping() || draining_all() || !started() || !can_hint_for(hid)) {
+        manager_logger.trace("Can't store a hint to {}", hid);
         ++_stats.dropped;
         return false;
     }
 
     try {
-        manager_logger.trace("Going to store a hint to {}", ep);
-        tracing::trace(tr_state, "Going to store a hint to {}", ep);
+        manager_logger.trace("Going to store a hint to {}", hid);
+        tracing::trace(tr_state, "Going to store a hint to {}", hid);
 
-        return get_ep_manager(ep).store_hint(std::move(s), std::move(fm), tr_state);
+        return get_host_manager(hid).store_hint(std::move(s), std::move(fm), tr_state);
     } catch (...) {
-        manager_logger.trace("Failed to store a hint to {}: {}", ep, std::current_exception());
-        tracing::trace(tr_state, "Failed to store a hint to {}: {}", ep, std::current_exception());
+        manager_logger.trace("Failed to store a hint to {}: {}", hid, std::current_exception());
+        tracing::trace(tr_state, "Failed to store a hint to {}: {}", hid, std::current_exception());
 
         ++_stats.errors;
         return false;
     }
 }
 
-bool manager::too_many_in_flight_hints_for(endpoint_id ep) const noexcept {
+bool manager::too_many_in_flight_hints_for(locator::host_id hid) const noexcept {
+    const auto ep = convert_host_id_to_ep(hid);
```

Comment:
We can't just use the current `token_metadata` from `storage_proxy`, it may have changed and doesn't contain host_id anymore.  We can use `is_me` version with `host_id`. With gossiper we could probably check that there is at least one endpoint among `get_nodes_with_host_id` which is `<= _max_hint_window_us`.